src/cargo/util/flock.rs

```diff
@@ -352,3 +342,153 @@ fn acquire(
         false
     }
 }
+
+#[cfg(unix)]
+mod sys {
+    use std::fs::File;
+    use std::io::{Error, Result};
+    use std::os::unix::io::AsRawFd;
+
+    pub(super) fn lock_shared(file: &File) -> Result<()> {
+        flock(file, libc::LOCK_SH)
+    }
+
+    pub(super) fn lock_exclusive(file: &File) -> Result<()> {
+        flock(file, libc::LOCK_EX)
+    }
+
+    pub(super) fn try_lock_shared(file: &File) -> Result<()> {
+        flock(file, libc::LOCK_SH | libc::LOCK_NB)
+    }
+
+    pub(super) fn try_lock_exclusive(file: &File) -> Result<()> {
+        flock(file, libc::LOCK_EX | libc::LOCK_NB)
+    }
+
+    pub(super) fn unlock(file: &File) -> Result<()> {
+        flock(file, libc::LOCK_UN)
+    }
+
+    pub(super) fn error_contended(err: &Error) -> bool {
+        err.raw_os_error().map_or(false, |x| x == libc::EWOULDBLOCK)
+    }
+
+    pub(super) fn error_unsupported(err: &Error) -> bool {
+        match err.raw_os_error() {
+            Some(libc::ENOTSUP) => true,
+            #[cfg(target_os = "linux")]
+            Some(libc::ENOSYS) => true,
+            _ => false,
+        }
+    }
+
+    #[cfg(not(target_os = "solaris"))]
+    fn flock(file: &File, flag: libc::c_int) -> Result<()> {
+        let ret = unsafe { libc::flock(file.as_raw_fd(), flag) };
+        if ret < 0 {
+            Err(Error::last_os_error())
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Simulate flock() using fcntl(); primarily for Oracle Solaris.
+    #[cfg(target_os = "solaris")]
+    fn flock(file: &File, flag: libc::c_int) -> Result<()> {
```

Comment:
I've got a few thoughts about this implementation. First off it seems like it's different enough from `flock` that it may want its own little submodule too?

Additionally I was [reviewing the differences](https://stackoverflow.com/questions/29611352/what-is-the-difference-between-locking-with-fcntl-and-flock) between fcntl vs flock-based locks. I don't think fcntl is what we want in terms of synchronization in that it releases the lock if *any* file description in the process closes the file or it also doesn't protect against threads. I don't *think* that Cargo runs afoul of this today but I don't think we can really guarantee that we won't ever run afoul of this ever.

If solaris doesn't implement the `flock`-style file locking, could solaris perhaps just eschew locking entirely?

Reply:
Yeah, the `fcntl`-style locking is an ugly hack.  In illumos, we ended up adding a "real" flock implementation to the OS.  Considering the existing bail-outs for NFS and other filesystems which don't tolerate locking, it's probably OK to use no-op locking for solaris targets.  I suspect that illumos systems awaiting their own `target_os` are probably the biggest pool of systems using the `solaris` target today.

Reply:
Yeah in general file locking is largely best-effort within Cargo and it's only relevant for concurrent Cargo invocations which are somewhat rare. Internally Cargo doesn't rely on file locks at all, so having it be a noop should keep everything largely working.

Reply:
Sounds great.  I'm happy to see that stuff go.